### PR TITLE
[Perf][Flux2][HunyuanVideo1.5] Skip attention-mask to avoid varlen path

### PR DESCRIPTION
**Pull request by kTorp**
## Purpose
This PR extends https://github.com/vllm-project/vllm-omni/pull/4293 and https://github.com/vllm-project/vllm-omni/pull/3763 to Flux2 and HunyuanVideo1.5 as per request. Similarly as before it improves latencies for sequence lengths not divisible by the world-size when running SP.

## Test Plan

- [x] Before and after results, with commands, shown below.

**vLLM Version: 0.23.0**

**vLLM-Omni Commit: ad07f98b7**

## Test Result
Commands used
```
VLLM_LOGGING_LEVEL=DEBUG vllm serve black-forest-labs/FLUX.2-dev \
--omni \
--port 8098 \
--ulysses-degree 8

curl -s -X POST http://localhost:8098/v1/images/generations \
-H "Content-Type: application/json" \
-d '{
    "prompt": "A small cat",
    "size": "1304x1304",
    "num_inference_steps": 25,
    "guidance_scale": 0.0,
    "seed": 42,
    "response_format": "b64_json"
}' \
| jq -r '.data[0].b64_json' \
| base64 -d > results/flux2_skip_mask_padded.png
```

```
VLLM_LOGGING_LEVEL=DEBUG vllm serve hunyuanvideo-community/HunyuanVideo-1.5-Diffusers-480p_t2v \
  --omni \
  --port 8098 \
  --ulysses-degree 8 \
  2>&1 | tee /app/results/hunyuan_serve.log

curl -sS -X POST http://localhost:8098/v1/videos/sync \
  -H "Accept: video/mp4" \
  -F "prompt=A cute cat" \
  -F "size=848x480" \
  -F "num_frames=33" \
  -F "fps=24" \
  -F "num_inference_steps=30" \
  -F "guidance_scale=6.0" \
  -F "flow_shift=5.0" \
  -F "seed=42" \
  -D /app/results/hunyuan_headers.txt \
  -o /app/results/hunyuanvideo_skip_mask_padded.mp4
```

| Model | Resolution / frames | SP padding | Baseline  | PR | decrease |
|-------|---------------------|-----------|--------------------------|----------------|---------|
| FLUX.2-dev | 1304×1304 | (6561→6568) | 7049.25 ms | 5444.68 ms | -22.8%|
| HunyuanVideo-1.5-T2V | 848×480, 33f | (14310→14312) | 16518.37 ms | 11729.65 ms | -29.0% |

Baseline Flux2 1304x1304

flux2_baseline_1304x1304

PR Flux2 1304x1304

flux2_skip_mask_1304x1304

Baseline HunyuanVideo1.5, 848×480, 33f

https://github.com/user-attachments/assets/45898bc0-b636-4446-898e-4260c1287823

PR HunyuanVideo1.5, 848×480, 33f

https://github.com/user-attachments/assets/236c616f-d1d1-4199-95ce-92d063aebf15

**BEFORE SUBMITTING:** read [CONTRIBUTING.md](https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md) and run the [precheck-pr skill](https://github.com/vllm-project/vllm-omni/blob/main/.claude/skills/precheck-pr/SKILL.md) with the code agent for a self-check against project conventions.
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by wtomin — APPROVED** [review]
LGTM.

**Comment by wtomin** [comment]
I notice there is no explanation of `mask_sp_padding` argument in `docs/user_guide/diffusion/parallelism/sequence_parallel.md`. Could you please add it ?

Besides,  `ulysses_mode="advanced_uaa"` is an experimental extension to Ulysses-SP. It lets Ulysses attention handle arbitrary sequence lengths and arbitrary attention head counts without relying on `attention_mask`-based token padding. Could you also compare `mask_sp_padding=False` and `ulysses_mode="advanced_uaa"`  in your PR?

**Comment by kTorp** [comment]
Thanks for review @wtomin 😄

I added a snippet to the md file, how does it look? Should I also add a bool flag to `entrypoints/serve.py` so it's settable during online serving as well?

I have only used uaa previously to handle divisibility of headcounts / devices, does it also handle padding/mask cases?

At least when testing flux2 with 1304x1304 resolution

```
VLLM_LOGGING_LEVEL=DEBUG vllm serve black-forest-labs/FLUX.2-dev \
--omni \
--port 8098 \
--ulysses-degree 8 \
--ulysses-mode advanced_uaa
```

logs still prints `Auto-padded sequence from 6561 to 6568 (pad_size=7, world_size=8, dim=1)` and has a similar 
latency `e2e_total_ms | 7,130.19` to the padded/masked branch.

**Comment by wtomin** [comment]
@kTorp I think the documentation is good.

As for the UAA behaviour：
1. FLUX.2-dev's _sp_plan hardcodes auto_pad=True in vllm_omni/diffusion/models/flux2/flux2_transformer.py:
```python
     _sp_plan = {
      "": {
          "hidden_states": SequenceParallelInput(split_dim=1, expected_dims=3, auto_pad=True),
      },
      "rope_prepare": {
          2: SequenceParallelInput(split_dim=0, expected_dims=2, split_output=True, auto_pad=True),
          3: SequenceParallelInput(split_dim=0, expected_dims=2, split_output=True, auto_pad=True),
      },
      "proj_out": SequenceParallelOutput(gather_dim=1, expected_dims=3),
     }
```
So the SP input hook pads hidden_states to make them divisible by `sequence_parallel_size` before Ulysses attention is entered.
  2. `advanced_uaa` works at a different layer UAA (in `vllm_omni/diffusion/attention/parallel/ulysses.py`) relaxes divisibility inside Ulysses attention:
    • It uses variable split sizes in `all_to_all_single(..., output_split_sizes=seq_lens)` so ranks can have different sequence lengths.
    • It pads/slices the head dimension when `head_cnt % ulysses_degree != 0`.
    • It does not use attention_mask-based token padding.
But because FLUX.2-dev already padded the tensors in the _sp_plan hook, UAA receives a padded, evenly-divisible sequence and basically falls back to near-even behavior.

I think it's alright that Flux-2 dev is hard-coded to support auto padding only. There is no need to test uaa performance for this model.

**Comment by wtomin** [comment]
I think it is safe to add this feature to skip attention mask for most sequence lengths (as the number of padded tokens is relatively small), but also cc @Bounty-hunter @david6666666 @gcanlin @RuixiangMa

**Comment by wtomin** [comment]
I searched `mask_sp_padding` key name in the current repo, only `wan2.2`, `wan2.2 vace`, and `qwen-image` support this key name.

Plus `flux-2` and `hunyuanvideo1.5`, there will be five pipelines supporting skip attention_mask. Please update this coverage in the doc.

**Comment by hsliuustc0106** [comment]
> I searched `mask_sp_padding` key name in the current repo, only `wan2.2`, `wan2.2 vace`, and `qwen-image` support this key name.
> 
> Plus `flux-2` and `hunyuanvideo1.5`, there will be five pipelines supporting skip attention_mask. Please update this coverage in the doc.

any updates needed?

**Comment by wtomin** [comment]
@kTorp Please update the related doc. Afterwards, I think this PR is good to be merged.

File: docs/user_guide/diffusion/parallelism/sequence_parallel.md
Comment (wtomin) [inline docs/user_guide/diffusion/parallelism/sequence_parallel.md:171]:
```suggestion
- Degrees must evenly divide the sequence length for optimal performance (or use `ulysses_mode="advanced_uaa"` for Ulysses-SP)
- `mask_sp_padding` is an experimental feature, currently only supported by `Wan2.2`, `Wan2.2 Vace`, `Qwen-Image`, `Flux 2`, and `HunyuanVideo 1.5`
```

**Comment by kTorp** [comment]
Hi @wtomin! 

Sorry for the lack of replies, I was away for holidays/vacation. 
Many thanks for updating the PR in my absence, much appreciated 🙏

**Comment by kTorp** [comment]
I've merged main branch and reran the experiments locally, similar performance improvements still observed and no quality degradation.
